### ISSUE TITLE
Get all descendants / ancestors

### ISSUE THREAD

**Issue by ORBAT** [facebook/ent#1024]
Hi folks! Is there any way to basically mimic what Postgres' [`WITH RECURSIVE`](https://www.postgresqltutorial.com/postgresql-recursive-query/) does, so that I could get eg. all followers' followers' followers' (etc etc) in the `m2mrecur` example https://github.com/facebook/ent/blob/9ff32ad60afd31c163cc76e7b208ebb627d7ee4f/examples/m2mrecur/ent/schema/user.go#L18-L32

So basically a way to find all reachable vertices from some vertex along some edge.

I'd be happy with just being able to provide a raw SQL expression, but I couldn't quite come up with a way to do that either.

**Comment by a8m** [comment]
Hey @ORBAT, the framework doesn't support this atm, and you need to do it explicitly as follows:

```go
client.User.
    Query().
    WithFollowers(func (q *ent.UserQuery) {
        q.WithFollowers()
    }).
    All(ctx)
```

We can leave this issue open for now, just to hear more thoughts, get suggestions for the API or even discuss if we want to support such thing or not - maybe it can be added as a plugin to ent.

**Comment by ORBAT** [comment]
Hi @a8m, and thanks for the quick reply.

The problem with your example is that it stops after one step, where I'm looking for a way to mimic `WITH RECURSIVE` that can go to any depth – I'm trying to find all leaf nodes of a certain type reachable from a node in a graph. Basically a breadth-first (or depth-first, don't really care) search. 

Doing it manually by iteratively fetching more nodes after each step is of course possible, but being able to use `WITH RECURSIVE` would make it muuuuch faster (plus less code too.) These sorts of graph traversals / searches are pretty common if you actually do have a graph data model, so it'd definitely be neat if `ent` supported them.

I'd definitely be perfectly happy with just _some_ way of getting the Postgres dialect to generate `WITH RECURSIVE` statements and then being able to use them somehow, doesn't have to be pretty.

Buuuuuut, thinking of possible APIs, I threw this together https://gist.github.com/ORBAT/9cc2bce522584ec45719d74ea969d628

**Comment by francismarcus** [comment]
My suggestions following up on @ORBAT gist

```go
	client.User.Query().
		Where(user.Country("fi")).
		Recursively().
		QuerySubordinates().
		Where(user.Country("fi"),
			user.OrganizationalUnit("r&d")).
		NoDuplicates().
		AllX(ctx)
```

**Comment by ORBAT** [comment]
@francismarcus, that `Recursively` modifier is actually much clearer than `Search*`, nice.

But yeah, I keep running into cases where this functionality would be very handy. I'm working on "porting" some old code over to using ent, and an actual example I have right now is figuring out which license a user is covered by. Users belong to groups, groups can have child / parent groups, and a group license covers that group and every group "under" it. It'd be handy to be able to do a recursive query to find the licenses in the group hierarchy

**Comment by francismarcus** [comment]
Yeah this would indeed be handy, if we add a Recursively modifier we should be able to run the already generated queries

**Comment by Naist4869** [comment]
+1

**Comment by nvthongswansea** [comment]
+1

**Comment by a8m** [comment]
I just want to update that I added initial support for `WITH RECURSIVE` statements in our SQL builder and you can see its usage in this example: https://github.com/ent/ent/blob/961dc08d148d3f00f8e1794830a86c253548c5d0/examples/fs/example_test.go#L53-L86

I hope to get more progress on it soon and add support for recursive traversals in the generated API.
